internal/types/document_validation.go

```diff
@@ -85,31 +89,60 @@ func (d *Document) ValidateData() error {
 		}
 		duplicateChecker[key] = struct{}{}
 
-		if key == "_id" {
+		if key == IdKey {
 			idPresent = true
 		}
 
 		value := must.NotFail(d.Get(key))
 
-		// TODO Add dance tests for infinity: https://github.com/FerretDB/FerretDB/issues/1151
-		if v, ok := value.(float64); ok && math.IsInf(v, 0) {
-			return newValidationError(
-				ErrValidation, fmt.Errorf("invalid value: { %q: %f } (infinity values are not allowed)", key, v),
-			)
+		var vErr *ValidationError
+
+		switch v := value.(type) {
+		case *Document:
+			err := v.ValidateData()
+			if err != nil && errors.As(err, &vErr) {
+				if vErr.code != ErrIDNotFound {
+					return err
+				}
+			}
```

Comment:
This will silently ignore error if `ValidateData()` ever in the future returns other than ValidationError. I know it's not the case now but I think we could be a bit more defensive here, WDYT?

Reply:
Nice catch! Fixed it.